source/server/filter_chain_manager_impl.h

```diff
@@ -234,34 +282,36 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   findFilterChainForSourceIpAndPort(const SourceIPsTrie& source_ips_trie,
                                     const Network::ConnectionSocket& socket) const;
 
+  const FilterChainManagerImpl* getOriginFilterChainManager() {
+    ASSERT(!add_filter_chain_done_);
+    return origin_;
+  }
+  // Duplicate the inherent factory context if any.
+  std::shared_ptr<Network::DrainableFilterChain>
+  findExistingFilterChain(const envoy::config::listener::v3::FilterChain& filter_chain_message);
+
+  // Mapping from filter chain message to filter chain. This is used by LDS response handler to
+  // detect the filter chains in the intersection of existing listener and new listener.
+  FcContextMap fc_contexts_;
+
   // Mapping of FilterChain's configured destination ports, IPs, server names, transport protocols
   // and application protocols, using structures defined above.
   DestinationPortsMap destination_ports_map_;
   const Network::Address::InstanceConstSharedPtr address_;
+  // This is the reference to a factory context which all the generations of listener share.
   Configuration::FactoryContext& parent_context_;
-  std::list<std::unique_ptr<Configuration::FilterChainFactoryContext>> factory_contexts_;
-};
+  std::list<std::shared_ptr<Configuration::FilterChainFactoryContext>> factory_contexts_;
 
-class FilterChainImpl : public Network::FilterChain {
-public:
-  FilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
-                  std::vector<Network::FilterFactoryCb>&& filters_factory)
-      : transport_socket_factory_(std::move(transport_socket_factory)),
-        filters_factory_(std::move(filters_factory)) {}
+  bool add_filter_chain_done_{false};
```

Comment:
Instead of an additional boolean I would use absl::optional if you need to differentiate between null and valid.

Reply:
Agree!  Fixed